File: Alert/app/src/main/java/org/alertpreparedness/platform/v1/min_preparedness/model/Action.java
package org.alertpreparedness.platform.v1.min_preparedness.model;

import android.net.Uri;

import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.Exclude;

import org.alertpreparedness.platform.v1.model.User;

/**
 * Created by faizmohideen on 13/12/2017.
 */

public class Action {
    private String id;
    private String assignee;
    private Long budget;
    private Long createdAt;
    private String department;
    private Long dueDate;
    private Boolean isComplete;
    private Boolean isArchived;
    private Boolean isInProgress;
    private String taskName;
    private String createdByAgencyId;
    private String createdByCountryId;
    private String networkId;
    @Exclude
    private Integer frequencyValue;
    private Long frequencyBase;
    private Long actionType;
    private Long level;
    private Long updatedAt;
    private boolean requireDoc;

    @Exclude
    private Uri path;

    @Exclude
    public User user;

    @Exclude
    public DatabaseReference db;

    @Exclude
    public DatabaseReference userRef;

    @Exclude
    public DatabaseReference networkRef;
    private boolean hasCHSInfo;
    private boolean isCHS;

    public Action() {
    }

    public Action(Uri path) {
        this.path = path;
    }

    public Action(Boolean isInProgress) {
        this.isInProgress = isInProgress;
    }

    public Action(String id, String taskName, String department, String assignee, String createdByAgencyId, String createdByCountryId, String networkId, Boolean isArchived, Boolean isComplete, Long createdAt, Long updatedAt,
                  Long actionType, Long dueDate, Long budget, Long level, Long frequencyBase, Integer frequencyValue, User user, DatabaseReference db, DatabaseReference userRef, DatabaseReference networkRef) {
        this.id = id;
        this.taskName = taskName;
        this.department = department;
        this.assignee = assignee;
        this.createdByAgencyId = createdByAgencyId;
        this.createdByCountryId = createdByCountryId;
        this.networkId = networkId;
        this.isArchived = isArchived;
        this.isComplete = isComplete;
        this.createdAt = createdAt;
        this.updatedAt = updatedAt;
        this.actionType = actionType;
        this.dueDate = dueDate;
        this.budget = budget;
        this.level = level;
        this.frequencyBase = frequencyBase;
        this.frequencyValue = frequencyValue;
        this.user = user;
        this.db = db;
        this.userRef = userRef;
        this.networkRef = networkRef;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getCreatedByAgencyId() {
        return createdByAgencyId;
    }

    public void setCreatedByAgencyId(String createdByAgencyId) {
        this.createdByAgencyId = createdByAgencyId;
    }

    public String getCreatedByCountryId() {
        return createdByCountryId;
    }

    public void setCreatedByCountryId(String createdByCountryId) {
        this.createdByCountryId = createdByCountryId;
    }

    public String getNetworkId() {
        return networkId;
    }

    public void setNetworkId(String networkId) {
        this.networkId = networkId;
    }

    public Boolean getInProgress() {
        return isInProgress;
    }

    public void setInProgress(Boolean inProgress) {
        isInProgress = inProgress;
    }

    @Exclude
    public Integer getFrequencyValue() {
        return frequencyValue;
    }

    @Exclude
    public void setFrequencyValue(Integer frequencyValue) {
        this.frequencyValue = frequencyValue;
    }

    public Long getFrequencyBase() {
        return frequencyBase;
    }

    public void setFrequencyBase(Long frequencyBase) {
        this.frequencyBase = frequencyBase;
    }

    public Long getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(Long createdAt) {
        this.createdAt = createdAt;
    }

    public Long getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(Long updatedAt) {
        this.updatedAt = updatedAt;
    }

    public String getTaskName() {
        return taskName;
    }

    public void setTaskName(String taskName) {
        this.taskName = taskName;
    }

    public String getDepartment() {
        return department;
    }

    public void setDepartment(String department) {
        this.department = department;
    }

    public String getAssignee() {
        return assignee;
    }

    public void setAssignee(String assignee) {
        this.assignee = assignee;
    }

    public long getActionType() {
        return actionType;
    }

    public void setActionType(Long actionType) {
        this.actionType = actionType;
    }

    public Long getDueDate() {
        return dueDate;
    }

    public void setDueDate(Long dueDate) {
        this.dueDate = dueDate;
    }

    public Long getBudget() {
        return budget;
    }

    public void setBudget(Long budget) {
        this.budget = budget;
    }

    public Boolean getArchived() {
        return isArchived;
    }

    public void setArchived(Boolean archived) {
        isArchived = archived;
    }

    public Boolean getComplete() {
        return isComplete;
    }

    public void setComplete(Boolean complete) {
        isComplete = complete;
    }

    public Long getLevel() {
        return level;
    }

    public void setLevel(Long level) {
        this.level = level;
    }

    public Uri getPath() {
        return path;
    }

    public void setPath(Uri path) {
        this.path = path;
    }


    @Override
    public String toString() {
        return "Action{" +
                "id='" + id + '\'' +
                ", isArchived=" + isArchived +
                ", getIsComplete=" + isComplete +
                ", isInProgress=" + isInProgress +
                ", taskName='" + taskName + '\'' +
                ", department='" + department + '\'' +
                ", assignee='" + assignee + '\'' +
                ", createdByAgencyId='" + createdByAgencyId + '\'' +
                ", createdByCountryId='" + createdByCountryId + '\'' +
                ", networkId='" + networkId + '\'' +
                ", frequencyValue=" + frequencyValue +
                ", frequencyBase=" + frequencyBase +
                ", actionLevel=" + actionType +
                ", dueDate=" + dueDate +
                ", budget=" + budget +
                ", level=" + level +
                ", createdAt=" + createdAt +
                ", updatedAt=" + updatedAt +
                ", ref=" + path +
                ", db=" + db +
                ", userRef=" + userRef +
                ", networkRef=" + networkRef +
                '}';
    }

    public boolean getRequireDoc() {
        return requireDoc;
    }

    public void setRequireDoc(boolean requireDoc) {
        this.requireDoc = requireDoc;
    }

    public void setHasCHSInfo(boolean hasCHSInfo) {
        this.hasCHSInfo = hasCHSInfo;
    }

    public boolean hasCHSInfo() {
        return hasCHSInfo;
    }

    public void setIsCHS(boolean isCHS) {
        this.isCHS = isCHS;
    }

    public boolean isCHS() {
        return isCHS;
    }
}

File: Alert/app/src/main/java/org/alertpreparedness/platform/v2/models/Action.kt
package org.alertpreparedness.platform.v2.models

import com.google.gson.annotations.SerializedName
import org.alertpreparedness.platform.v2.models.enums.ActionLevel
import org.alertpreparedness.platform.v2.models.enums.ActionType
import org.joda.time.DateTime

class Action(
        @SerializedName("asignee")
        val assignee: String?,
        val budget: Double?,
        val createdAt: DateTime,
        @SerializedName("department")
        val departmentId: String,
        val dueDate: DateTime?,
        val isComplete: Boolean,
        val isCompleteAt: DateTime?,
        @SerializedName("level")
        val actionLevel: ActionLevel,
        @SerializedName("type")
        val actionType: ActionType,
        val requireDoc: Boolean,
        val task: String,
        val updatedAt: DateTime,
        val isArchived: Boolean
) : BaseModel() {

    @Transient
    lateinit var clockSettings: ClockSettings

    fun getExpirationTime(): DateTime {
        return if (isCompleteAt != null) {
            isCompleteAt.plus(clockSettings.calculateOffset())
        } else {
            updatedAt.plus(clockSettings.calculateOffset())
        }
    }

    override fun toString(): String {
        return "Action(assignee='$assignee', budget=$budget, createdAt=$createdAt, departmentId='$departmentId', dueDate=$dueDate, isComplete=$isComplete, isCompleteAt=$isCompleteAt, actionLevel=$actionLevel, requireDoc=$requireDoc, task='$task', type=$actionType, updatedAt=$updatedAt)"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Action) return false

        if (assignee != other.assignee) return false
        if (budget != other.budget) return false
        if (createdAt != other.createdAt) return false
        if (departmentId != other.departmentId) return false
        if (dueDate != other.dueDate) return false
        if (isComplete != other.isComplete) return false
        if (isCompleteAt != other.isCompleteAt) return false
        if (actionLevel != other.actionLevel) return false
        if (actionType != other.actionType) return false
        if (requireDoc != other.requireDoc) return false
        if (task != other.task) return false
        if (updatedAt != other.updatedAt) return false
        if (isArchived != other.isArchived) return false
        if (clockSettings != other.clockSettings) return false

        return true
    }

    override fun hashCode(): Int {
        var result = assignee?.hashCode() ?: 0
        result = 31 * result + (budget?.hashCode() ?: 0)
        result = 31 * result + createdAt.hashCode()
        result = 31 * result + departmentId.hashCode()
        result = 31 * result + (dueDate?.hashCode() ?: 0)
        result = 31 * result + isComplete.hashCode()
        result = 31 * result + (isCompleteAt?.hashCode() ?: 0)
        result = 31 * result + actionLevel.hashCode()
        result = 31 * result + actionType.hashCode()
        result = 31 * result + requireDoc.hashCode()
        result = 31 * result + task.hashCode()
        result = 31 * result + updatedAt.hashCode()
        result = 31 * result + isArchived.hashCode()
        result = 31 * result + clockSettings.hashCode()
        return result
    }
}